book-content/chapters/07-mutability.md

```diff
@@ -0,0 +1,1013 @@
+
+#### Adding a Type to the Object
+
+Another option is to explicitly declare the type of the `albumAttributes` object to be `AlbumAttributes`:
+
+```typescript
+const albumAttributes: AlbumAttributes = {
+  status: "on-sale",
+};
+
+updateStatus(albumAttributes); // No error
+```
+
+This works similarly to how it did with the `let`. While `albumAttributes.status` can still be reassigned, it can only be reassigned to a valid value:
+
+```typescript
+albumAttributes.status = "new-release"; // No error
+```
+
+This behaviour works the same for all object-like structures, including arrays and tuples. We'll examine those later in the exercises.
+
+### Readonly Object Properties
+
+In JavaScript, as we've seen, object properties are mutable by default. But TypeScript lets us be more specific about whether or not a property of an object can be mutated.
+
+To make a property read-only (not writable), you can use the `readonly` modifier:
+
+Consider this `Album` interface, where the `title` and `artist` are marked as `readonly`:
+
+```typescript
+interface Album {
+  readonly title: string;
+  readonly artist: string;
+  status?: "new-release" | "on-sale" | "staff-pick";
+  genre?: string[];
+}
+```
+
+Once an `Album` object is created, its `title` and `artist` properties are locked in and cannot be changed. However, the optional `status` and `genre` properties can still be modified.
+
+Note that this only occurs on the _type_ level. At runtime, the properties are still mutable. TypeScript is just helping us catch potential errors.
+
+#### The `Readonly` Type Helper
+
+If you want to specify that _all_ properties of an object should be read-only, TypeScript provides a type helper called `Readonly`.
+
+To use it, you simply wrap the object type with `Readonly`.
+
+Here's an example of using `Readonly` to create an `Album` object:
+
+```typescript
+const readOnlyWhiteAlbum: Readonly<Album> = {
+  title: "The Beatles (White Album)",
+  artist: "The Beatles",
+  status: "staff-pick",
+};
+```
+
+Because the `readOnlyWhiteAlbum` object was created using the `Readonly` type helper, none of the properties can be modified:
+
+```ts twoslash
+// @errors: 2540
+type Album = {
+  title: string;
+  artist: string;
+  status?: "new-release" | "on-sale" | "staff-pick";
+  genre?: string[];
+};
+
+const readOnlyWhiteAlbum: Readonly<Album> = {
+  title: "The Beatles (White Album)",
+  artist: "The Beatles",
+  status: "staff-pick",
+};
+// ---cut---
+readOnlyWhiteAlbum.genre = ["rock", "pop", "unclassifiable"];
+```
+
+Note that like many of TypeScript's type helpers, the immutability enforced by `Readonly` only operates on the first level. It won't make properties read-only recursively.
+
+### Readonly Arrays
+
+As with object properties, arrays and tuples can also be made immutable by using the `readonly` modifier.
```

Comment:
i never thought about it before but it's interesting that `readonly` modifier for arrays/tuples is applied on the whole thing - making all elements readonly - while in objects we can apply that only on property level